### PR TITLE
[OptApp] Improving terminal log output 

### PR DESCRIPTION
**Pull request by RezaNajian**
This is a quick PR that enhances the logging of OptApp, particularly at the algorithm level.
here is a screenshot

image

### CODE REVIEW

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py [hunk 1]
```diff
@@ -101,22 +102,23 @@ def Solve(self):
                         algorithm_data.GetBufferedData()["abs_obj[%]"] = self.__objective.GetAbsoluteChange() / initial_value * 100
                     print(self.__objective.GetInfo())
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py]:
```suggestion
                    Kratos.Logger.PrintInfo("AlgorithmSteepestDescent", "\n" + self.__objective.GetInfo())
```

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py [hunk 2]
```diff
@@ -101,22 +102,23 @@ def Solve(self):
...
-                with TimeLogger("Calculate design update", None, "Finished"):
+                with TimeLogger("AlgorithmSteepestDescent::Calculate design update", None, "Finished"):
                     search_direction = self.ComputeSearchDirection(obj_grad)
                     algorithm_data.GetBufferedData()["search_direction"] = search_direction
-
                     alpha = self.__line_search_method.ComputeStep()
-
                     update = search_direction * alpha
                     self.__control_field += update
+                    print(self.__line_search_method.GetInfo())
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py]:
```suggestion
                    Kratos.Logger.PrintInfo("AlgorithmSteepestDescent", "\n" + self.__line_search_method.GetInfo())
```
                    Kratos.Logger.PrintInfo("AlgorithmSteepestDescent", "\n" + self.__objective.GetInfo())

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py [hunk 3]
```diff
@@ -101,22 +102,23 @@ def Solve(self):
...
 
                 algorithm_data.GetBufferedData()["parameter_update"] = update
                 algorithm_data.GetBufferedData()["control_field"] = self.__control_field
 
-                self.converged = self.__convergence_criteria.IsConverged()
+                with TimeLogger("AlgorithmSteepestDescent::Check convergence", None, "Finished"):
+                    self.converged = self.__convergence_criteria.IsConverged()
+                    print(self.__convergence_criteria.GetInfo())
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py]:
```suggestion
                    Kratos.Logger.PrintInfo("AlgorithmSteepestDescent", "\n" + self.__convergence_criteria.GetInfo())
```

**Review by sunethwarna — DISMISSED** [review]
This is really cool. Could you may be put a sample output in the PR to see. :D

**Comment by RezaNajian** [comment]
@sunethwarna ready for your review

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_constraint.py [hunk 4]
```diff
@@ -140,14 +146,15 @@ def GetRelativeChange(self) -> float:
...
\ No newline at end of file
+    def GetInfo(self) -> dict:
+        info = {
+            "name": self.GetReponse().GetName(),
+            "value": self.GetValue(),
+            "type": self.__constraint_type,
+            "ref_value": self.GetStandardizedReferenceValue(),
+            "abs_change": self.GetAbsoluteChange(),
+            "rel_change [%]": self.GetRelativeChange() * 100.0,
+            "abs_violation [%]": self.GetAbsoluteViolation(),
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_constraint.py]:
```suggestion
            "abs_violation [%]": self.GetAbsoluteViolation() * 100,
```

**Review by Igarizza — DISMISSED** [review]
Nice work

**Review by Igarizza — DISMISSED** [review]
sure

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py [hunk 5]
```diff
@@ -115,11 +123,13 @@ def Solve(self):
...
                 alpha = self.__line_search_method.ComputeStep()
 
-                self.UpdateControlField(alpha)
+                self.ComputeControlUpdate(alpha)
 
-                self.converged = self.__convergence_criteria.IsConverged()
+                self.Output()
 
-                self.CallOnAllProcesses(["output_processes"], Kratos.OutputProcess.PrintOutput)
+                self.UpdateControl()
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py:130]: Yes, I see, in old version the control_field was from next iteration (after adding update). Good catch, thnx!

**Review by Igarizza — APPROVED** [review]
good catch

**Review by Igarizza — CHANGES_REQUESTED** [review]
@RezaNajian @sunethwarna Let's use decorator instead of with TimeLogger. Here is an example: 
https://www.geeksforgeeks.org/timing-functions-with-decorators-python/

It would allow to do the same thing as you do but without one Tab level and you need to add only one decorator line to use it and all your implementation can be fully reused.  

I will try it and post here.

**Comment by RezaNajian** [comment]
@Igarizza Thanks for the point. I have checked the link and also discussed it with Suneth. That improves and saves the implementation. I would suggest extending the current implementation in a separate PR.

**Review by Igarizza — APPROVED** [review]
I have crearted a following PR with decorator #11326 . You can merge.

**Pull request merged**